src/bt.rs

```diff
@@ -74,324 +73,217 @@ pub enum BinaryTreeError<V> {
-                });
-
-                stack.push(Item {
-                    name: format!("{prefix}left:"),
-                    level: level + 1,
-                    node: left.as_deref(),
-                });
-            } else {
-                writeln!(f, "{name} <None>")?
-            };
-        }
+    fn encoded_len(&self) -> Option<usize> {
+        self.value
+            .encoded_len()
+            .zip(self.left.encoded_len())
+            .zip(self.right.encoded_len())
+            .map(|((v, l), r)| v + l + r)
+    }
+}
 
-        Ok(())
+impl<V: Decode> Decode for Node<V> {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            value: V::decode(bytes)?,
+            left: Option::<usize>::decode(bytes)?,
+            right: Option::<usize>::decode(bytes)?,
+        })
     }
 }
 
 impl<V: Encode> Encode for BinaryTree<V> {
     fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
-        match &self.root {
-            None => CodecMarker::Leaf.encode(bytes),
-            Some(node) => {
-                CodecMarker::Inner.encode(bytes)?;
-                node.encode(bytes)
-            }
+        let len = self.nodes.len();
+        UsizeType::try_from(len)
+            .map_err(|_| CodecError::UnexpectedValue)?
+            .encode(bytes)?;
+
+        for node in self.nodes.iter() {
+            node.encode(bytes)?;
         }
-    }
-}
 
-impl<P, V: ParameterizedDecode<P>> ParameterizedDecode<P> for BinaryTree<V> {
-    fn decode_with_param(param: &P, bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
-        Ok(match CodecMarker::decode(bytes)? {
-            CodecMarker::Leaf => Self::default(),
-            CodecMarker::Inner => Self {
-                root: Some(Box::new(Node::decode_with_param(param, bytes)?)),
-            },
-        })
+        Ok(())
     }
-}
 
-#[cfg(feature = "test-util")]
-impl<V> core::fmt::Display for BinaryTree<V>
-where
-    V: core::fmt::Display,
-    Node<V>: core::fmt::Display,
-{
-    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "---")?;
-        if let Some(node) = &self.root {
-            writeln!(f, "{node}")?
-        };
-        Ok(())
+    fn encoded_len(&self) -> Option<usize> {
+        UsizeType::encoded_len(&0)
+            .zip(self.nodes.first().map_or(Some(0), Encode::encoded_len))
+            .map(|(header_len, node_len)| header_len + node_len * self.nodes.len())
     }
 }
 
-/// Marker used to distinguish between full and empty nodes.
-#[repr(u8)]
-#[derive(Debug)]
-enum CodecMarker {
-    Leaf,
-    Inner,
+impl<V: Decode> Decode for BinaryTree<V> {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let mut tree = Self::default();
+        let header = UsizeType::decode(bytes)?;
+        if header != 0 {
+            let len = header.into();
+            tree.nodes = Vec::with_capacity(len + Self::NODES_CAPACITY);
```

Comment:
nit: instead of allocating a new vector (which will toss the vector of size `NODES_CAPACITY` allocated in the `Default` implementation), `reserve` `len` additional elements. This will perform no allocation if the storage space is large enough already, and a minimal allocation if it is not.

Alternatively, drop this line entirely, and let `push` resize the vector appropriately.

Reply:
good catch.